public/ocr/ocr_main.js

```diff
@@ -844,6 +863,291 @@ async function resetDevices() {
+TILE_ID_TO_NTC_BLOCK_ID.set(0x7d, 2);
+TILE_ID_TO_NTC_BLOCK_ID.set(0x7c, 3);
+
+let data_frame_buffer = new Uint8Array(GAME_FRAME_SIZE);
+
+async function initCaptureFromEverdrive() {
+	frame_duration = 1000 / config.frame_rate;
+	const ports = await navigator.serial.getPorts();
+	if (ports.length) {
+		everdrive = ports.find(port => {
+			const { usbProductId, usbVendorId } = port.getInfo();
+			return (
+				usbVendorId === EVERDRIVE_N8_PRO.usbVendorId &&
+				usbProductId === EVERDRIVE_N8_PRO.usbProductId
+			);
+		});
+
+		if (everdrive) {
+			captureFromEverdrive();
+			return;
+		}
+	}
+
+	everdrive = await navigator.serial.requestPort({
+		filters: [EVERDRIVE_N8_PRO],
+	});
+
+	if (everdrive) {
+		captureFromEverdrive();
+	}
+}
+
+async function captureFromEverdrive() {
+	try {
+		await everdrive.open({ baudRate: 115200, bufferSize: GAME_FRAME_SIZE }); // plenty of speed for 60fps data frame from gym are 132 bytes: 132x60=7920
+	} catch (err) {
+		console.warn(err);
+		// assume port is already open for now
+		// TODO: better error checking
+	}
+
+	everdrive_reader = everdrive.readable.getReader({ mode: 'byob' });
+	everdrive_writer = everdrive.writable.getWriter();
+
+	// verify we have a real everdrive by sending a GET_STATUS command (expecting [0x00, 0xA5] as response)
+	const bytes = [
+		'+'.charCodeAt(0),
+		'+'.charCodeAt(0) ^ 0xff,
+		0x10, // getStatus
+		0x10 ^ 0xff,
+	];
+
+	await everdrive_writer.write(new Uint8Array(bytes));
+	const { value, done } = await everdrive_reader.read(data_frame_buffer);
+
+	if (value[0] !== 0 || value[1] !== 0xa5) {
+		console.error('Selected device is not an everdrive');
+		return; // How to recover?
+	}
+
+	// restore the buffer for next use
+	data_frame_buffer = new Uint8Array(value.buffer);
+
+	console.log('Everdrive verified!');
+
+	start_time = Date.now();
+
+	requestFrameFromEverDrive();
+}
+
+function convertTwoBytesToDecimal(byte1, byte2) {
+	return byte2 * 100 + (byte1 >> 4) * 10 + (byte1 & 0xf);
+}
+
+/*
+function updateField(field, tetriminoX, tetriminoY, currentPieceOrientation) {
+
+}
+/**/
+
+async function requestFrameFromEverDrive() {
+	performance.mark('edlink_comm_start');
+
+	// 0. prep request
+	// ref: https://github.com/zohassadar/EDN8-PRO/blob/nestrischamps/edlink-n8/edlink-n8/Edio.cs#L622
+	const bytes = [
+		// command header
+		'+'.charCodeAt(0),
+		'+'.charCodeAt(0) ^ 0xff,
+		EVERDRIVE_CMD_MEM_WR,
+		EVERDRIVE_CMD_MEM_WR ^ 0xff,
+
+		// addr
+		EVERDRIVE_ADDR_FIFO & 0xff,
+		(EVERDRIVE_ADDR_FIFO >> 8) & 0xff,
+		(EVERDRIVE_ADDR_FIFO >> 16) & 0xff,
+		(EVERDRIVE_ADDR_FIFO >> 24) & 0xff,
+
+		// len
+		1,
+		0,
+		0,
+		0,
+
+		// exec
+		0,
+
+		EVERDRIVE_CMD_SEND_STATS,
+	];
+
+	// 1. send request
+	const res = await everdrive_writer.write(new Uint8Array(bytes));
+
+	console.log(res);
+
+	performance.mark('edlink_write_end');
+	console.log('sent stats command');
+
+	// 2. read response
+	const { value, done } = await everdrive_reader.read(data_frame_buffer);
```

Comment:
What appears to be happening is that sometimes a partial frame is returned in `value`.  When that happens the next read will begin where the last one ended.   

Here is how I was able to work around:  https://github.com/zohassadar/nestrischamps/commit/27ac00678ae770abe0ac51f0863bcbec76310d62 (fair warning: I do not know how to javascript)

Reply:
Ah right, thanks for that!

My code was clearly bad 🤦 . I assumed that in `byob` mode, the buffer would automatically get filled before it was returned, but seeing as there is still a `done` flag, that should been am obvious clue that it is not, duh 😑 

I'll tweak your patch to make things work while maintaining the `byob` mode.

I found a nice example of getting it done here:
https://developer.chrome.com/en/articles/serial/#read-port

```javascript
async function readInto(reader, buffer) {
  let offset = 0;
  while (offset < buffer.byteLength) {
    const { value, done } = await reader.read(
      new Uint8Array(buffer, offset)
    );
    if (done) {
      break;
    }
    buffer = value.buffer;
    offset += value.byteLength;
  }
  return buffer;
}

const reader = port.readable.getReader({ mode: "byob" });
let buffer = new ArrayBuffer(512);
// Read the first 512 bytes.
buffer = await readInto(reader, buffer);
// Then read the next 512 bytes.
buffer = await readInto(reader, buffer);
```